drivers: ethernet: xlnx_gem: switch integrated MDIO/PHY support over to Zephyr's framework

ibirnbaum opened the pull request:
When the device driver for the GEM Ethernet controller was first implemented, Zephyr's framework for MDIO and PHY devices was still in its very early stages and was too limited for use in conjunction with the GEM MAC. While MDIO-less use cases exist for the GEM (such as connecting it to an Ethernet switch in the FPGA part of the chip), pretty much every Zynq-/ZynqMP-based eval board comes with a PHY, and in order to support more flexible use than just supporting the one default speed setting configured by the FSBL at power-up, including the basic ability to detect link up/down, MDIO and PHY support were integrated directly into the driver. The GEM driver was one of two drivers to do it this way, and this internal implementation outlasted the appearance of generic MDIO and PHY support for quite some time.

Now that the limitations of the early stages of the generic MDIO/PHY support within Zephyr no longer exist, I propose switching the GEM's MDIO and PHY capabilities over to the facilities Zephyr provides nowadays. This to me appears beneficial as the addition of a new supported PHY (such as the TI DP83867 which is common in conjunction with the ZynqMP) is not limited to a single Ethernet MAC and also, it's not an addition to code that will inevitably at some point be a dead end if demands are made for the remaining device drivers bypassing the dedicated frameworks to switch over.

Considering that we're talking about two supported SoCs, one in-tree target board definition with a supported (simulated) PHY, the MDIO driver itself, two supported PHYs and a lot of code that has to be shifted around and broken up into pieces as well as obsolete configuration items from the DT that are removed due to them being replaced with those provided by the MDIO/PHY framework, this PR is a pretty big package. However, it contains everything that is required to just keep running what was already running - the integration into the MDIO shell is one little feature that is added on top.

There's no consideration yet if there's settings that could or should be added to the supported PHYs' DTs (a few PHYs have custom DT items, such as the DP83867's "RX strap quirk" setting), or if any setting that is being applied in the PHY drivers' code should be enabled by default as they are right now (e.g. should 'robust auto MDIX' always be on by default in the TI PHY).

This PR would open up the use of the PHYs so far only supported by the GEM to everyone, allowing anyone who is more familiar with those PHYs and thinks that things should be done differently to make suggestions in that direction.

ibirnbaum committed:
drivers: ethernet: xlnx_gem: remove legacy MDIO/PHY code

Remove the legacy MDIO and PHY support for the GEM MAC which
dates back to 2021 and doesn't use the MDIO/PHY frameworks
provided by Zephyr nowadays, as those didn't exist back then.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
drivers: mdio: add MDIO driver for Xilinx GEM

Separate the MDIO functionality from the Xilinx GEM MAC driver
into a separate driver conforming with the standard MDIO API.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
dts: bindings: xlnx_gem: remove legacy MDIO-related properties

Remove any properties from the Xilinx GEM DT node specification and the
associated header in dt-bindings that relate to the legacy MDIO/PHY
support.

Consistently remove all instances of these properties from the DTs of the
Zynq-7000 and ZynqMP SoCs and from the DT of the qemu_cortex_a9 board.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
dts: zynq7000: fix AMBA AHB burst, DMA packet buffer configuration

Fix configuration of GEM0, GEM1:
- AMBA AHB burst length changed from SINGLE to INCR16
- RX/TX packet buffers in the DMA now fit a full Ethernet frame

This prevents DMA errors and issues with frames that are spread across
multiple buffers.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
drivers: ethernet: xlnx_gem: switch over to new MDIO driver

Remove all data/functions/function calls relating to the now
removed legacy MDIO/PHY support directly integrated into the
GEM device driver. Switch over to using the standard MDIO/PHY
interfaces which will now refer to the new separate MDIO
driver and the separated PHY drivers.

Consistently update the qemu_cortex_a9 board definition with
the matching MDIO and PHY nodes, qemu_cortex_a9 is currently
the only Zynq-based target where an enabled GEM instance is
specified by default.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
dts: zynqmp: fix AMBA AHB burst, DMA packet buffer configuration

Fix configuration of GEM0, GEM1:
- AMBA AHB burst length changed from SINGLE to INCR16
- RX/TX packet buffers in the DMA now fit a full Ethernet frame

This prevents DMA errors and issues with frames that are spread across
multiple buffers.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

pdgendt commented on drivers/ethernet/phy/phy_marvell_88e1xxx.c:
```diff
@@ -0,0 +1,624 @@
+	return 0;
+}
+
+static DEVICE_API(ethphy, marvell_alaska_phy_api) = {.get_link = phy_marvell_alaska_get_link,
+						     .cfg_link = phy_marvell_alaska_cfg_link,
+						     .link_cb_set = phy_marvell_alaska_link_cb_set,
+						     .read = phy_marvell_alaska_read,
+						     .write = phy_marvell_alaska_write};
```
Nit, put a trailing comma after the last entry so it can be formatted like: (same for other driver)

```c
static DEVICE_API(ethphy, marvell_alaska_phy_api) = {
	.get_link = phy_marvell_alaska_get_link,
	.cfg_link = phy_marvell_alaska_cfg_link,
	.link_cb_set = phy_marvell_alaska_link_cb_set,
	.read = phy_marvell_alaska_read,
	.write = phy_marvell_alaska_write,
};
```

ibirnbaum replied:
done, thanks

pdgendt replied:
Not fully yet, move the function after `{` to a new line with a single indent.

ibirnbaum replied:
I changed the formatting as per your request, however: if I didn't botch the install of the relevant tooling, the original formatting should have been a direct result of Zephyr's own clang-format file. Maybe there's some changes to be made there?

maass-hamburg commented:
@ibirnbaum I saw that you added two phy drivers, what is the benefit of them, to the already existing generic ``ethernet-phy``? We actually use the DP83822 in one of our products with the generic ethernet phy driver and it just works. 
Using for example the interrupt of a phy instead of polling the link state would be such a benefit.

maass-hamburg reviewed (changes requested):
entry in the migration guide needed

ibirnbaum commented:
> @ibirnbaum I saw that you added two phy drivers, what is the benefit of them, to the already existing generic `ethernet-phy`? We actually use the DP83822 in one of our products with the generic ethernet phy driver and it just works. Using for example the interrupt of a phy instead of polling the link state would be such a benefit.

I can add the interrupt support, however, as mentioned in the PR's description, my initial intention was just to carry over what was already working in the driver's local MDIO/PHY code to Zephyr's respective frameworks and then go for extensions afterwards. If combining such a transitional PR with new features isn't an issue for you, I can add the support now. What's your preference?

When it comes to having a separate driver aside from the generic one, there is some selfishness involved. The reason is that Weidmüller uses the DP83822 (and its predecessor, the TLK105) in conjunction with several industrial field buses which use Ethernet as physical layer (e.g. EtherCAT, Ethernet/IP, Profinet, Powerlink, Modbus TCP) in the field bus couplers of the u-remote product range.

For EtherCAT, ProfiNet and Ethernet/IP, some settings which go beyond the generic register set have to be altered. Depending on which bus is implemented, those settings relate to LED behavior (MLEDCR register), Analog Power Detect Control (APDS/APDC registers), Auto MDIX, Force MDIX, Robust Auto MDIX (PHYCR/CR1 registers) and Fast Link Down (CR3 register). My intention would be to make those settings available to everyone, likely in the form of DT properties. The old PHY code custom to the GEM driver contained some of those (and they'd be carried over for now with the adoption of this PR), but we so far haven't gone public with everything we have nowadays as a) our exact settings probably won't work for someone else, b) we knew that maintaining the custom GEM MDIO/PHY code would likely be a dead end eventually, and c) the varying configurations per firmware build variant are probably better handled via the DT rather than littering the driver code with ifdefs. The assumption is that those options for this PHY might eventually also be useful to someone else.

The afforementioned DP83867 GBit PHY will need a separate driver as it has a specific issue revolving around the bootstrap configuration pins. Certain boards exist, amongst them UltraScale boards made by Xilinx themselves (e.g. ZCU102), where the bootstrap configuration sometimes causes the PHY to come up in self-test mode, which has to be handled in software. In Linux, a custom DT property exists for this issue, IIRC named something like "rx-strap-quirk".

Regarding the Marvell Alaska PHYs, I'll check to what extent they're compatible with the generic approach. Configuration of those PHYs comes with some bank switching and frequent resets, and at least the common Digilent-made Zynq-based eval boards I know of couldn't be used for implementing interrupt support as the IRQ line simply isn't wired up on those boards. If the PHY's configuration approach requires the custom driver, somebody else will have to retrofit the interrupt support.

ibirnbaum commented:
> entry in the migration guide needed

Noted, I'll take care of it

maass-hamburg commented:
> > @ibirnbaum I saw that you added two phy drivers, what is the benefit of them, to the already existing generic `ethernet-phy`? We actually use the DP83822 in one of our products with the generic ethernet phy driver and it just works. Using for example the interrupt of a phy instead of polling the link state would be such a benefit.
> 
> I can add the interrupt support, however, as mentioned in the PR's description, my initial intention was just to carry over what was already working in the driver's local MDIO/PHY code to Zephyr's respective frameworks and then go for extensions afterwards. If combining such a transitional PR with new features isn't an issue for you, I can add the support now. What's your preference?
> 
> When it comes to having a separate driver aside from the generic one, there is some selfishness involved. The reason is that Weidmüller uses the DP83822 (and its predecessor, the TLK105) in conjunction with several industrial field buses which use Ethernet as physical layer (e.g. EtherCAT, Ethernet/IP, Profinet, Powerlink, Modbus TCP) in the field bus couplers of the u-remote product range.
> 
> For EtherCAT, ProfiNet and Ethernet/IP, some settings which go beyond the generic register set have to be altered. Depending on which bus is implemented, those settings relate to LED behavior (MLEDCR register), Analog Power Detect Control (APDS/APDC registers), Auto MDIX, Force MDIX, Robust Auto MDIX (PHYCR/CR1 registers) and Fast Link Down (CR3 register). My intention would be to make those settings available to everyone, likely in the form of DT properties. The old PHY code custom to the GEM driver contained some of those (and they'd be carried over for now with the adoption of this PR), but we so far haven't gone public with everything we have nowadays as a) our exact settings probably won't work for someone else, b) we knew that maintaining the custom GEM MDIO/PHY code would likely be a dead end eventually, and c) the varying configurations per firmware build variant are probably better handled via the DT rather than littering the driver code with ifdefs. The assumption is that those options for this PHY might eventually also be useful to someone else.
> 
> The afforementioned DP83867 GBit PHY requires a separate driver as it has a specific issue revolving around the bootstrap configuration pins. Certain boards exist, amongst them UltraScale boards made by Xilinx themselves (e.g. ZCU102), where the bootstrap configuration sometimes causes the PHY to come up in self-test mode, which has to be handled in software. In Linux, a custom DT property exists for this issue, IIRC named something like "rx-strap-quirk". I'll check in the existing driver if this is already handled properly.
> 
> Regarding the Marvell Alaska PHYs, I'll check to what extent they're compatible with the generic approach. Configuration of those PHYs comes with some bank switching and frequent resets, and at least the common Digilent-made Zynq-based eval boards I know of couldn't be used for implementing interrupt support as the IRQ line simply isn't wired up on those boards. If the PHY's configuration approach requires the custom driver, somebody else will have to retrofit the interrupt support.

If you can get both phys working with just the generic phy driver, its preferred to use that now and and the more specific phys in a later PR, when the extra features can be set and used.

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -647,17 +620,25 @@ static enum ethernet_hw_caps eth_xlnx_gem_get_capabilities(
-		caps |= ETHERNET_DUPLEX_SET;
-	}
-
 	if (dev_conf->copy_all_frames) {
 		caps |= ETHERNET_PROMISC_MODE;
 	}
 
```
similar with ``ETHERNET_PROMISC_MODE``, this should tell, if ETHERNET_CONFIG_TYPE_PROMISC_MODE is supported in set_config

ibirnbaum replied:
Resolved via the other PR that introduced the set_config function.

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -225,21 +217,22 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
 	net_if_set_link_addr(iface, dev_data->mac_addr, 6, NET_LINK_ETHERNET);
 	ethernet_init(iface);
-	net_if_carrier_off(iface);
+	net_eth_carrier_off(iface);
```
consider making the ``phy-handle`` required in the ``xlnx,gem.yaml``. It seems to be needed to configure the GEM Network Configuration Register and clock speed. also without a phy the carrier would never be set to on.

ibirnbaum replied:
done

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -225,21 +217,22 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
 	const struct device *dev = net_if_get_device(iface);
 	const struct eth_xlnx_gem_dev_cfg *dev_conf = dev->config;
 	struct eth_xlnx_gem_dev_data *dev_data = dev->data;
+#ifdef CONFIG_MDIO
```
remove all ``#ifdef CONFIG_MDIO``, the phy could be on i2c or spi.

ibirnbaum replied:
done

maass-hamburg commented on boards/qemu/cortex_a9/Kconfig.defconfig:
```diff
@@ -29,6 +29,9 @@ if NETWORKING
 config NET_L2_ETHERNET
 	default y
 
+config MDIO
```
maybe add ``select MDIO if  DT_HAS_XLNX_GEM_MDIO_ENABLED`` to ``menuconfig ETH_XLNX_GEM`` in drivers/ethernet/Kconfig.xlnx_gem instead. could also be a imply instead of a select

ibirnbaum replied:
done

ibirnbaum commented:
@pdgendt @maass-hamburg 
Sorry for the delay on this PR, will get back to all pending comments shortly.

ibirnbaum commented:
Re-based against main due to the introduction of set_config in a separate PR. Some of the comments have been taken care of already, others are still to follow. Checking that the tests still pass after the rebase. Haven't gotten around to testing the DP83822 and the Alaska PHYs with just the generic PHY support, that's still on my list.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -998,52 +957,52 @@ static void eth_xlnx_gem_set_initial_nwcfg(const struct device *dev)
 		reg_val |= ETH_XLNX_GEM_NWCFG_1000_BIT;
 	}
+	/* Set FDEN bit for full-duplex operation */
+	if (state->speed & (LINK_FULL_10BASE_T | LINK_FULL_100BASE_T |
```
done

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -620,24 +607,10 @@ static enum ethernet_hw_caps eth_xlnx_gem_get_capabilities(
+	caps |= ETHERNET_LINK_1000BASE_T |
+		ETHERNET_LINK_100BASE_T |
+		ETHERNET_LINK_10BASE_T |
+		ETHERNET_DUPLEX_SET;
```
I see, removed the flag for now, now that there's a set_config function, I can add manual duplex / link / auto-neg capabilities later on.

ibirnbaum commented:
@maass-hamburg Regarding your comments on:

- removing the MDIO device pointer from the driver
- making the phy-dev handle required

The GEM controller as found in the Xilinx SoCs has some properties / design aspects that differ from other FPGA-less SoCs incorporating the GEM, which is originally a Cadence IP used by Xilinx among others, which I think result in neither of your suggestions being feasible. The big issue is that the assumption 'the respective controller instance is connected to the outside world via a managed PHY, because it's capable of doing so', be it via the controller's own MDIO interface, or, as you mentioned, via I2C/SPI (actually a setup I haven't worked with yet) isn't universally valid in the Xilinx SoCs.

It's not uncommon with the Xilinx SoCs to configure one or more GEM(s) to not 'face the outside world' via MIO pins in the form of an RGMII interface with a PHY attached, but 'face inwards' to the FPGA part of the SoC in the form of a GMII interface, which can, for example, be hooked up to an FPGA IP implementing a [n]-Port Ethernet Switch (which in turn can expose its ports either internally or externally, eventually potentially leading to the need for MDIO facilities for a PHY not associated with an Ethernet controller - but that's another issue for some other time). In such a scenario, neither MDIO nor a PHY are involved for the respective controller instance, and its link properties don't ever have to be altered at run-time. Yet, at the same time, one or more other GEM instance(s) (1 in the Zynq, up to 3 in the ZynqMP) could still 'face outwards' with MDIO and PHY enabled.

Such system designs would be prevented if a PHY device handle was universally mandatory in conjunction with the GEM controller, therefore, I'd rather not make phy-handle mandatory.

Regarding the MDIO device pointer:
From my point of view, this is required for two reasons:
1) support of the GMII-to-RGMII FPGA IP
2) only mapping the respective GEM's register space once in MMU-based systems once the driver is switched over to the DEVICE_MMIO API

Re 1): The Xilinx SoCs have two different types of I/O pins: MIO pins, which are associated with the Processor System - the part of the SoC the GEM controllers reside in - and EMIO pins, which are accessible only from within the FPGA part of the SoC. When a significant number of the Processor System's integrated peripherals are to be connected to the outside world, it's not uncommon, in particular with the smaller chip packages, to not have enough MIO pins available to expose them all directly. In that case, the workaround is to just route their interface signals into the FPGA part of the SoC and just 'hardwire' them to available EMIO pins.

When using this workaround for the GEM Ethernet controllers, the media interface automatically changes from RGMII to GMII the moment the respective controller's interface signals are routed into the FPGA part. In order to end up with an RGMII interface via EMIO pins just as you would when using MIO pins, the system design must include the GMII-to-RGMII Converter IP in the FPGA. This converter becomes a separate node on the respective GEM controller's MDIO bus in addition to the external PHY, usually at address 8. At least older versions of this IP consist of just a single register into which the detected link speed has to be written whenever a link speed change is detected by the PHY (more recent versions snoop the MDIO bus for reads of the relevant status registers from the PHY and extract the newly detected link speed automatically).

Whenever the GMII-to-RGMII Converter IP is in use, it not only converts interface signals, but also takes the generation of the TX clock away from the controller. Therefore, what it uses the link speed information (either written explicitly or snooped from the MDIO bus) for is to generate a TX clock matching either 10, 100 or 1000 MBps. Which means that, when an older version of this IP core is in use, and I'd like to support that eventually, the driver's eth_xlnx_gem_configure_clocks() function must not calculate matching clock dividers for the master PLL driving the controller, but perform an MDIO write operation to a node that is not the PHY, just passing on the detected link speed for the FPGA-based clock generator to adjust to. In case of newer versions of the converter IP, it wouldn't technically have to do anything, however, as the newer IP versions are backwards compatible, I'd like to keep the implementation identical.

Support for the separate MDIO node that is the optional GMII-to-RGMII Converter is my primary motivation for the existence of the MDIO device pointer in the GEM driver's local data. But then, there's also...

Re 2): I've occasionally switched over drivers used in conjunction with the MMU-equipped Zynq-7000 to the DEVICE_MMIO API whenever possible in order to eventually get rid of the respective entries in the SoC's static MMU mapping table, which are #ifdef'd depending on the respective devices being enabled in the DT. This is where I'd eventually like to go with the GEM driver as well.

However, there's a bit of an issue there: the MDIO interface is a subset of the respective controller's register space. I.e., if the MDIO driver has its own DEVICE_MMIO mapping and the GEM driver has its own DEVICE_MMIO mapping, two MMU entries with different virtual addresses will point to the same physical address that is the respective controller's 4k register space. While this technically works, I'd rather not have that even if the code for the two drivers comes from the same person and mappings based on DEVICE_MMIO entries always happen the same way. This is due to ARM's statement in the ARM Architecture Reference Manual for ARMv7-A:

`If a single physical memory location has more than one set of attributes assigned to it, ARM strongly
recommends that software ensures that the sets of attributes are identical.`

The ARMv7 MMU code at the time being has no safeguards against what ARM calls 'mismatched memory attributes'. Therefore, my idea regarding automated mapping of the register space just once is:
- both the MDIO driver and the GEM driver are equipped with DEVICE_MMIO_ROM and DEVICE_MMIO_RAM entries.
- If the respective GEM uses MDIO, the MDIO driver maps the register space due to its higher init priority and the GEM driver acquires the mapped virtual address from the associated MDIO device - hence the MDIO device pointer in the GEM's data struct.
- If the respective GEM doesn't use MDIO, the GEM driver performs the mapping itself using its own DEVICE_MMIO facilities.

This admittedly isn't fully tested yet, as this to me is a future feature that's not in the scope of this MDIO-/PHY-related PR. It's a concept that I'd like to try out for the reason mentioned above, but this is secondary to the GMII-to-RGMII Converter support I'm going to need the MDIO device - not the PHY device - for.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -258,8 +220,30 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
+	if (dev_conf->phy_dev != NULL) {
+		net_eth_carrier_off(iface);
+
+		ret = phy_link_callback_set(dev_conf->phy_dev, eth_xlnx_gem_phy_cb,
```
I've re-worked things quite a bit once more, including a change of the DT structure (mdio is now a child node of the MAC, which prevents faulty nodelabel-based resolving of the MDIO node if both GEM0 and GEM1 are enabled, GEM1's MDIO is enabled but GEM0's is not, in that case, the one existing node associated with GEM1 will be labeled gem0_mdio). A phy-handle is no longer required, the PHY is now obtained via an array of the MDIO node's child devices. This also lays the groundwork for the afforementioned GMII-to-RGMII converter, which can later be implemented as a PHY device. If there is a converter instance, the MDIO node will have two child devices, if it's just a PHY, it'll be just one.

I've confirmed that the Marvell PHYs which can be found on some Digilent boards work with the generic PHY driver as long as no special features are requested. Therefore, also based on your statement that the DP83822 works with the generic drivers as well, both additional PHY drivers are out for now.

I'm currently working on link configuration via the set_config function, then I'll have a look at how the framework handles fixed links. The big push should be there in a few days.

ibirnbaum replied:
I eventually went back to using the phy-handle. Comp. the qemu_cortex_a9 DT.

maass-hamburg commented on drivers/mdio/mdio_xlnx_gem.c:
```diff
@@ -0,0 +1,302 @@
+
+#define XLNX_GEM_MDIO_DEV_CONFIG(port)                                                             \
+	static const struct xlnx_gem_mdio_config xlnx_gem##port##_mdio_cfg = {                     \
+		.gem_base_addr = DT_REG_ADDR_BY_IDX(DT_INST(port, xlnx_gem), 0),                   \
```
if it needs the base address of the ethernet mac, it's a better way to have a common parent and the ethernet mac and mdio as a child of it. like here: 
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/ethernet/st%2Cstm32-ethernet-controller.yaml and 
https://github.com/zephyrproject-rtos/zephyr/blob/179045e5a570e5a6e978651ecd3c4e3f7a3a4a4e/dts/arm/st/f7/stm32f765.dtsi#L71-L94

michalsimek commented:
Please rebase and I will approve.

michalsimek commented on dts/arm/xilinx/zynqmp.dtsi:
```diff
@@ -125,7 +122,13 @@
-			full-duplex;
+		};
+
+		gem0_mdio: gem0_mdio {
```
In Linux world DT node names shouldn't use _ but should use - instead. 
Also see this error message when used.
/tmp/kv260_r5/zephyr/zephyr.dts:149.24-154.5: Warning (simple_bus_reg): /soc/gem0_mdio: missing or empty reg/ranges property

ibirnbaum replied:
Zephyr's DT generation tools produce a syntax error when using dashes instead of underscores.

michalsimek commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -90,51 +96,18 @@ DT_INST_FOREACH_STATUS_OKAY(ETH_XLNX_GEM_INITIALIZE)
  */
 static int eth_xlnx_gem_dev_init(const struct device *dev)
 {
 	const struct eth_xlnx_gem_dev_cfg *dev_conf = dev->config;
```
__maybe_unused  when you removed if (dev_conf->init_phy) { below? 

.../zephyr-workspace/zephyr/drivers/ethernet/eth_xlnx_gem.c:109:44: warning: unused variable 'dev_conf' [-Wunused-variable]
  109 |         const struct eth_xlnx_gem_dev_cfg *dev_conf = dev->config;
      |                                            ^~~~~~~~

ibirnbaum replied:
done, added an attribute unused marker to suppress the warning

maass-hamburg commented on drivers/ethernet/phy/Kconfig:
```diff
@@ -96,6 +96,22 @@ config PHY_QUALCOMM_AR8031
 	help
 	  Enable Qualcomm Atheros AR8031 Ethernet PHY Driver
 
+config PHY_MARVELL_ALASKA_88E1XXX
+	bool "Marvell Alaska 88E1xxx PHY Driver"
+	default y
+	depends on DT_HAS_MARVELL_88E1XXX_ENABLED
+	depends on MDIO
+	help
+	  Enable Marvell Alaska 88E1xxx PHY Driver
+
+config PHY_TI_DP83822
+	bool "TI DP83822 PHY Driver"
+	default y
+	depends on DT_HAS_TI_DP83822_ENABLED
+	depends on MDIO
+	help
+	  Enable TI DP83822 PHY Driver
+
```
select MDIO

ibirnbaum replied:
resolved, outdated as those drivers are gone for now

ibirnbaum commented:
> Hi @ibirnbaum
> 
> The MDIO/PHY framework pull request has been open for long time and needs a rebase. Do you have any plans to push the updated changes addressing the review comments soon?
> 
> We’re planning to add additional features for the GEM driver on top of this framework, but we’re currently blocked as the PR is still open.
> 
> To move things forward, I was thinking to create a new pull request that incorporates the review comments and push the changes with your sign-off details. However, I’m not sure if that’s the best approach, so I’d like to hear your opinion before proceeding.
> 
> @michalsimek @HariniKatakamX

Yes, I'm working on this, however, as I'm doing so as a private individual, it's mostly weekend work only. I've re-based locally, but that still doesn't change the unfortunate fact that the build-time device class priority assertions and the capabilities of Zephyr's DT_... macros make this a pain to pull off in a half-decent fashion when it comes to the DT layout, also considering that there might be, aside from the PHY, another optional node for the GMII-to-RGMII-converter. It looks like we'll end up with custom device class init priorities for any SoC family supporting the GEM, which isn't great, but there's precedence for this on main.

Please consider that there's also still the PR which moves the DMA to nocache memory and also makes the driver ZynqMP compatible. Whichever PR gets approved and merged first will cause another rebase for the other. From my point of view, further changes which go beyond a hand full of lines of code should probably eventually be based on the driver including the re-worked DMA and MDIO.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -414,21 +382,22 @@ ETH_NET_DEVICE_DT_INST_DEFINE(port,\
 /* Device configuration data declaration macro */
 #define ETH_XLNX_GEM_DEV_CONFIG(port) \
 static const struct eth_xlnx_gem_dev_cfg eth_xlnx_gem##port##_dev_cfg = {\
+	.mdio_dev			= ETH_XLNX_GEM_MDIO_DEV(port),\
```
done, I think the planned GMII-to-RGMII-converter support can also be pulled off without this

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -680,18 +607,14 @@ struct eth_xlnx_gem_bdring {
  * UltraScale SoCs, which both contain the GEM.
  */
 struct eth_xlnx_gem_dev_cfg {
+	const struct device		*mdio_dev;
```
done

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -529,7 +490,7 @@ static void eth_xlnx_gem##port##_irq_config(const struct device *dev)\
 /* RX/TX BD Ring initialization macro */
 #define ETH_XLNX_GEM_INIT_BD_RING(port) \
-if (dev_conf->base_addr == DT_REG_ADDR_BY_IDX(DT_INST(port, xlnx_gem), 0)) {\
+if (dev_conf->base_addr == DT_REG_ADDR_BY_IDX(DT_PARENT(DT_INST(port, xlnx_gem)), 0)) {\
```
here too

maass-hamburg commented on drivers/ethernet/CMakeLists.txt:
```diff
@@ -9,15 +9,11 @@ zephyr_library_sources_ifdef(CONFIG_ETH_DRIVER_RAW_MODE
-	phy_xlnx_gem.c
-	)
+  eth_gecko.c
+  phy_gecko.c
```
these indentation changes seem unrelated to this pr, please change

maass-hamburg reviewed (changes requested):
please retain the bi-selectability (https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements) for each commit. If you do breaking changes, the fix should be in the same commit. It should be working even between the commits.

venodela commented:
Hi @ibirnbaum 

Could you please Rebase this PR?

The following PR is now merged:
drivers: ethernet: xlnx_gem: fix packet DMA configuration on ZynqMP
https://github.com/zephyrproject-rtos/zephyr/pull/95695

@HariniKatakamX

venodela commented:
Thanks for the update, @ibirnbaum.

This PR is a dependency for some of our ongoing development work.

We’ll wait for the rebased version.

@@HariniKatakamX

ibirnbaum commented:
> > > Hi @ibirnbaum
> > > DNM label can be removed for this Pull Request;
> > 
> > 
> > This rebase is a dry run so far, as I don't have access to actual hardware for a final test run before Thursday. Also, I'm currently re-basing some more extensions of the driver against this code base.
> 
> ok. FYI; With previous version of pull request i verified on AMD Versal Gen2 R52 hardware. it worked for me;

Sounds good! I'll take care of the lower end of things, i.e. the Zynq-7000 and the ZynqMP RPU on the ZCU102. By doing so, the tests will cover both an MMU-based as well as an MPU-based system.

ibirnbaum commented:
@maaikez I've done some testing and debugging over the past two nights (sorry for getting side-tracked to the folks who are actually waiting on the MDIO side of things, that's still coming up, it's 1:30 AM by now so that's not going to happen tonight any more), and I hope the results may be helpful.

For starters, yes, the performance on the Zynq was never stellar, this was already the case when the DMA was still located in the OCM with the previous version of the driver. Interestingly, my test results were better on the ZynqMP RPU, which runs at a lower CPU clock frequency (by the way, my specs when it comes to this comparison and also to comparisons to your particular board are: for the Zynq, default Zedboard configuration, CPU 666 MHz, DDR 533 MHz, for the ZynqMP, default ZCU102 configuration, CPU 500 MHz, DDR also 533 MHz). Whether this is solely down to the differences between DDR3 and DDR4 RAM, or the caches (which are 32k L1 I/D on both systems), or MMU vs. MPU behaviour, I can't immediately tell. In both cases, we're comparing 32-bit CPUs here, so 64-bit doesn't come into play as a performance factor directly, although the ZynqMP's GEM has a 64-bit data bus width even when used by the RPU vs. the Zynq's 32-bit DBUS width.

There's not a lot I can say about the FreeRTOS/lwip combo, I only had to deal with that once on the ZynqMP RPU and it wasn't by choice, also, Ethernet throughput performance didn't matter for the use case this setup covered. What might be a general performance consideration is that contrary to other implementations (I've been told this at least about Linux's network stack), Zephyr's stack and its interface to the Ethernet drivers probably copy data around more than others do. You can't just provide a pointer to an RX buffer in the DMA to the network stack when processing received data, drivers always have to allocate packet buffers and copy the data at least once from the DMA to the buffer before it likely gets copied again into the destination socket RX buffer. I guess performance would be noticeably better if it wasn't for copying received data in the driver. Also, with the default configuration, RX'ed buffers are being handed over from the driver to the network stack in the context of the system workqueue, whose priority you might try adjusting. Setting the attribute `handle-rx-in-isr;` in your respective GEM's device tree node changes that, but with little performance gains as far as I can tell. Also, depending on the traffic load, it makes the execution time of the ISR less predictable.

Regarding the errors you're seeing:

- If you don't have the MMU enabled, then you also don't have the L1 caches enabled. Also, the Zynq's L2 cache, although manageable via CMSIS' API, isn't usable in conjunction with Zephyr and enabling it will cause the system to crash. All of the standard stuff in the Zynq is covered by the SoC-level MMU setup, although it is important that the modification of soc/xlnx/zynq7000/common/soc.c contained in this PR is also applied on your system. Do you per chance have any AXI IPs integrated that are not setting themselves up in the MMU? What exactly happens when you have the MMU enabled, do you get a data abort exception? If so, check the DFAR register value and compare the given address to your setup's memory map.

Also, without the MMU, I'm kinda surprised how far you've gotten with testing, I remember having to add the initial MMU support for aarch32 way back just because of the GEM, as unaligned memory access only works with the MMU enabled and this kind of access does happen when poking around in network packets...

- The error message 'unexpected missing SOF bit...' has so far occurred whenever packets are received at high frequency which span more than one buffer in the RX DMA. In the past I believed this to be a cache problem, as when debugging it I kept logging the control words encountered by the driver in a buffer and whenever this error occurred dumped them out. At some point (unless with artificial delays introduced when iterating over the buffer descriptors), the driver would always read control words from the buffer descriptors that didn't make sense, but what the driver saw was the direct result of a sys_read32 operation, so the issue must have been more on the hardware side of things. I didn't encounter this right now with my setup with MMU and caches enabled and the buffer descriptors being located in nocache memory (which I double-checked, the MMU processed the section placement directive from the driver code correctly). However, a general workaround is to set in your board's dts file, under the gem0/gem1 node, the values
`rx-buffer-size = <1536>;`
`tx-buffer-size = <1536>;`
Changing the RX/TX buffer count for the DMA doesn't seem to really improve performance either. In the end, the more packet buffers there are in cached memory, the more likely it is that nothing is really properly cacheable without getting displaced frequently due to the small L1 cache sizes, I guess.

- Regarding bit [10] in the ISR register, i.e. error indication 0x400: what made this go away was setting a DMA configuration item from the reference configuration (Zynq TRM chapter 16.3.2, instructions for the DMA control register) that I had previously set at my board level during development and testing and which so far still has a different default value in the zynq7000.dtsi: set your respective GEM's `amba-ahb-burst-length` property to `<XLNX_GEM_AMBA_AHB_BURST_INCR16>;`. Without this setting (default is SINGLE instead), I could reproduce that the controller kept throwing this error under load even with the caches disabled, so it wasn't a matter of proper nocache allocation of the BD rings and also not of proper cache invalidation management for the payload data.

I'd say that the most important thing for you to get up and running for reliable test results would be to get the MMU and with it the caches up and running. I'll have to see to changing the burst parameter by default in the zynq7000.dtsi so that it doesn't have to be changed for every single board...

ibirnbaum commented:
Resolved the merge conflict on the migration guide, MDIO driver now supports clause 45 as mentioned before. Default DMA configuration parameters (RX/TX buffer sizes, AHB burst length) are now set in the DT to the values stated in the Zynq TRM.

I'm not seeing any activity from the CI right now - checking back tomorrow morning. Once that's all clear, the DNM will go.

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -257,8 +219,22 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
+			return;
+		}
+	} else {
+		net_eth_carrier_on(iface);
```
the carrier is on by default, no need for this

ibirnbaum replied:
done, removed

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -366,8 +342,7 @@ static int eth_xlnx_gem_send(const struct device *dev, struct net_pkt *pkt)
 
-	if (!dev_data->started || dev_data->eff_link_speed == LINK_DOWN ||
-			(!net_if_flag_is_set(dev_data->iface, NET_IF_UP))) {
+	if (!dev_data->started || !net_if_flag_is_set(dev_data->iface, NET_IF_UP)) {
```
I don't really remember the details of this behaviour any more, however, back in the old days, the network stack would attempt to TX some IPv6 related stuff pretty early on, when the capability to send wasn't there yet. I can't remember if this was mitigated by the started check or the interface status check. If the 'up' flag is now considered at a higher level, perfect. I removed that part of the check.

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1777,3 +1574,32 @@ static void eth_xlnx_gem_handle_tx_done(const struct device *dev)
+
+	if (dev_data->iface == NULL) {
+		return;
+	}
```
done, removed

ibirnbaum commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -257,8 +219,22 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
+	 * For PHY-less operation: fixed link is pre-configured, indicate link up.
+	 */
+	if (dev_conf->phy_dev != NULL) {
+		net_eth_carrier_off(iface);
```
done, changed

maaikez commented:
So, everything is passed and done, can it be merged now? What are we waiting for actually?

BTW, using this driver for some time now, works fine.

maass-hamburg commented:
> So, everything is passed and done, can it be merged now? What are we waiting for actually?
> 
> BTW, using this driver for some time now, works fine.

2 approvals and zephyr no longer being in feature freeze

maass-hamburg commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -609,6 +609,15 @@ Ethernet
 * Ethernet drivers no longer need to call :c:func:`net_if_set_link_addr` themselves, when
   the mac address is changed via ``ETHERNET_CONFIG_TYPE_MAC_ADDRESS``. (:github:`105931`)
 
+* The Xilinx GEM Ethernet driver (:dtcompatible:`xlnx,gem`) has been switched over to use the
```
needs to be moved to the 4.5 file

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -435,21 +404,17 @@ ETH_NET_DEVICE_DT_INST_DEFINE(port,\
 	&eth_xlnx_gem_apis,\
 	NET_ETH_MTU);
 
+#define ETH_XLNX_GEM_PHY_DEV(port)\
+	DEVICE_DT_GET_OR_NULL(DT_INST_PHANDLE(port, phy_handle))
+
```
t seems like some bits need to be set depending on the current speed, therefore the phy_handle should not be optional. (people could still use the fixed link phy)

ibirnbaum replied:
I'll look into that, but the Xilinx/AMD SoCs definitely need the option to run this controller completely without MDIO and (actual hardware) PHY for things like Ethernet Switch IP cores in the FPGA - this is pretty regular stuff and not niche stuff like the weird checksum behaviour of some industrial field buses I sometimes come up with. I guess a fixed link setup should do the trick here, however, I haven't tested that by now.

ibirnbaum replied:
all fine regarding fixed link as substitute for an actual PHY, phy-handle is now required, "_OR_NULL" stripped from PHY device resolver macro.

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -366,8 +340,7 @@ static int eth_xlnx_gem_send(const struct device *dev, struct net_pkt *pkt)
 
-	if (!dev_data->started || dev_data->eff_link_speed == LINK_DOWN ||
-			(!net_if_flag_is_set(dev_data->iface, NET_IF_UP))) {
+	if (!dev_data->started) {
```
`dev_data->started` should not be needed, the networking subsystem already takes care, that we don't send anything before starting it.

ibirnbaum replied:
removed, done

venodela commented:
@ibirnbaum, could you please help us know your plan to resolve the open comments?

venodela commented:
@ibirnbaum Looking for your response on this Pull request.

maass-hamburg commented on soc/xlnx/zynq7000/common/soc.c:
instead the mmio api should be used instead https://github.com/zephyrproject-rtos/zephyr/pull/107244

ibirnbaum replied:
Is that possible considering that this is a separate device configuration register space aside from the controller's register space? I was under the impression that it has previously only been possible to map a single address space using DEVICE_MMIO.

ibirnbaum replied:
OK, it seems like DEVICE_MMIO_NAMED_ROM could be a viable alternative. I'll look into that.

However, an unfortunate side effect of either approach seems to be that at least on the Zynq, we'll end up with more than one page table entry mapping the SLCR register space (as the Zynq syscon driver, on top of which a pinctrl driver exists, does so as well), which is kinda discouraged by ARM AFAIK. But considering that the clean alternative is likely implementing full pin control and clock management for all Xilinx/AMD SoCs and integrate the use of those facilities into every single driver supported on them in order to defer access to the SLCR fully and cleanly to the respective pin control/clock management drivers, two page table entries for the same physical address are likely the less painful solution.

maass-hamburg commented on dts/arm/xilinx/zynq7000.dtsi:
```diff
@@ -49,55 +49,69 @@
+				unicast-hash;
+			};
+
+			gem0_mdio: mdio {
+				compatible = "xlnx,gem-mdio";
+				status = "disabled";
+				#address-cells = <1>;
+				#size-cells = <0>;
+			};
```
you can "now" put the mdio controller inside the ethernet mac node, xlnx,gem-controller is no longer needed

ibirnbaum replied:
done

ibirnbaum commented:
Just did a rebase against current main locally, and at least as far as git is concerned, there were no major collisions there.

So at the time being, I'm putting on my list of things to do:
- DEVICE_MMIO support
- Re-structuring the DT to place the MDIO node within the main GEM node
- Release notes: move from 4.4 to 4.5
- checking if the changes from #106086 affect PHY handling.

ibirnbaum committed:
dts: zynq7000: add DEVICE_MMIO compatibility

Add reg-names property for each GEM controller to enable compatibility with
the DEVICE_MMIO facilities.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
dts: zynqmp: dts: add DEVICE_MMIO compatibility

Add reg-names property for each GEM controller to enable compatibility with
the DEVICE_MMIO facilities.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum commented:
Current status as of tonight is:
- MDIO nodes moved below MAC nodes in SoC DTs
- removed outer DT node grouping MAC and MDIO in the SoC DTs
- DEVICE_MMIO is now integrated


- migration guide not yet switched to new version
- no review yet if anything's missing / incorrect due to the recent addition of the interface pointer to the driver API functions 
- NOT YET TESTED ON HARDWARE - more on that coming up

maass-hamburg commented on drivers/ethernet/mdio/mdio_xlnx_gem.c:
```diff
@@ -0,0 +1,336 @@
+ * This struct contains all device configuration data for a GEM
+ * MDIO interface instance which is constant.
+ */
+struct xlnx_gem_mdio_dev_cfg {
```
you only need a reference to the parent just like in https://github.com/zephyrproject-rtos/zephyr/pull/108046

ibirnbaum replied:
changed & tested, done

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -114,49 +121,17 @@ DT_INST_FOREACH_STATUS_OKAY(ETH_XLNX_GEM_BUFFER_SIZE_CHECK)
 {
-	const struct eth_xlnx_gem_dev_cfg *dev_conf = dev->config;
-	uint32_t reg_val;
+	const struct eth_xlnx_gem_dev_cfg *dev_conf __attribute__((unused)) = DEV_CFG(dev);
```
use `__maybe_unused`

ibirnbaum replied:
done

venodela commented:
Hi @ibirnbaum 

Could you please help us know your progress and pending activities on this PR?

venodela commented:
Hi @ibirnbaum

Could you please help us know your progress and pending activities on this PR?

There are several dependencies on this PR from my side, and it is currently blocking my work on other feature enhancements.

Please help me know if you need any help testing it on hardware.

ibirnbaum commented:
> Hi @ibirnbaum
> 
> Could you please help us know your progress and pending activities on this PR?
> 
> There are several dependencies on this PR from my side, and it is currently blocking my work on other feature enhancements.
> 
> Please help me know if you need any help testing it on hardware.

It's currently the holiday season of Ascension and Pentecost, so I've been on vacation for the past 1 1/2 weeks and will be back on Tuesday. Last thing I did prior to that was to verify functionality of the current state on the Zynq, where the DEVICE_MMIO facilities actually have an effect as opposed to the Cortex-R cores on the ZynqMP. ZynqMP RPU tests are still pending, as well as making sure that fixed link specification works as well so that use of the GEM interfacing with the PL with no MDIO and no PHY involved is possible.

venodela commented:
> > Hi @ibirnbaum
> > Could you please help us know your progress and pending activities on this PR?
> > There are several dependencies on this PR from my side, and it is currently blocking my work on other feature enhancements.
> > Please help me know if you need any help testing it on hardware.
> 
> It's currently the holiday season of Ascension and Pentecost, so I've been on vacation for the past 1 1/2 weeks and will be back on Tuesday. Last thing I did prior to that was to verify functionality of the current state on the Zynq, where the DEVICE_MMIO facilities actually have an effect as opposed to the Cortex-R cores on the ZynqMP. ZynqMP RPU tests are still pending, as well as making sure that fixed link specification works as well so that use of the GEM interfacing with the PL with no MDIO and no PHY involved is possible.

Thank you, @ibirnbaum, for the updates.

Since this PR has been open for a year, as it is blocking other driver enhancements.

 if needed any help in resolving open comments and if you need any help on testing it on hardware with newly commited changes. i'm happy to do that. please help me know?

and could you consider planning to add the fixed link support in a separate PR? 

@vineethchowdhary @HariniKatakamX

ibirnbaum commented:
> > > Hi @ibirnbaum
> > > Could you please help us know your progress and pending activities on this PR?
> > > There are several dependencies on this PR from my side, and it is currently blocking my work on other feature enhancements.
> > > Please help me know if you need any help testing it on hardware.
> > 
> > 
> > It's currently the holiday season of Ascension and Pentecost, so I've been on vacation for the past 1 1/2 weeks and will be back on Tuesday. Last thing I did prior to that was to verify functionality of the current state on the Zynq, where the DEVICE_MMIO facilities actually have an effect as opposed to the Cortex-R cores on the ZynqMP. ZynqMP RPU tests are still pending, as well as making sure that fixed link specification works as well so that use of the GEM interfacing with the PL with no MDIO and no PHY involved is possible.
> 
> Thank you, @ibirnbaum, for the updates.
> 
> Since this PR has been open for a year, as it is blocking other driver enhancements.
> 
> if needed any help in resolving open comments and if you need any help on testing it on hardware with newly commited changes. i'm happy to do that. please help me know?
> 
> and could you consider planning to add the fixed link support in a separate PR?
> 
> @vineethchowdhary @HariniKatakamX

Regarding the fixed link, it's not a matter of developing a new functionality, it's rather a matter of testing an existing functionality that has to be configured to match the use case with some supported target board.

maass-hamburg commented:
> and could you consider planning to add the fixed link support in a separate PR?

If you need fixed link support, use the `ethernet-phy-fixed-link` phy (`zephyr/drivers/ethernet/phy/phy_fixed_link.c`). it is exactly for that purpose

ibirnbaum committed:
dts: bindings: xlnx_gem: declare phy-handle property required

As PHY-less use cases such as the GEM being connected to an
Ethernet switch IP core in the FPGA part of the SoC can be
facilitated by declaring a fixed link instead of an actual PHY
in the target board's DT, declare the phy-handle property as
required in the GEM MAC's DT binding in order to prevent
undefined link states and incomplete link speed configuration
in the NWCFG register.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
doc: releases: migration guide entry for release 4.5

Add the migration guide entry for release 4.5 currently
in development.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

cfriedt commented:
@ibirnbaum - is it possible to reorder commits to ensure bisectability?

By the look of it, removing item from DTS first would prevent that.

I would suggest adding new items to DTS first, making the switch and then removing old items after.

venodela commented:
Hi @ibirnbaum,
  
  Thanks for the latest push. I reviewed the PR.
  
  Following up on @cfriedt's bisectability concern.

  Commits 1/2 (8b42aca, 9be4489) remove mdio-phy-address, link-speed, mdc-divider, phy-poll-interval and full-duplex from the DTSI, but dts/bindings/ethernet/xlnx,gem.yaml still declares those as required: true
  until commit 10 (e682879). Any bisect checkout between commits 1–9 will fail DevicetreeBindings validation because required properties are missing from the DTSI.
  
  @HariniKatakamX

maass-hamburg commented:
it is probably best to put the board, driver and dts changes in one commit

ibirnbaum committed:
dts: zynq7000: add MDIO nodes below MAC nodes

Add the MDIO nodes below the GEM0 and GEM1 nodes now that the corresponding
compatible is available.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum committed:
dts: zynqmp: add MDIO nodes below MAC nodes

Add the MDIO nodes below the GEM0/1/2/3 nodes now that the corresponding
compatible is available.

Signed-off-by: Immo Birnbaum <mail@birnbaum.immo>

ibirnbaum commented:
@maass-hamburg I'm having issues with running QEMU for qemu_cortex_a9 with networking mode set to Ethernet, has your recent switch from using the -nic parameter to using -netdev in conjunction with -device in https://github.com/zephyrproject-rtos/zephyr/commit/d692931947feacc9ae53d82ca774c36f2f7fc2ea been verified to work with the Xilinx QEMU fork contained in the SDK? I'm currently on SDK 1.0.1.

maass-hamburg commented:
> @maass-hamburg I'm having issues with running QEMU for qemu_cortex_a9 with networking mode set to Ethernet, has your recent switch from using the -nic parameter to using -netdev in conjunction with -device in [d692931](https://github.com/zephyrproject-rtos/zephyr/commit/d692931947feacc9ae53d82ca774c36f2f7fc2ea) been verified to work with the Xilinx QEMU fork contained in the SDK? I'm currently on SDK 1.0.1.

No, i have never used the Xilinx QEMU fork.

ibirnbaum commented:
@cfriedt I've re-arranged the commits and moved individual aspects from one commit to another, is the whole add/remove sequence the way it should be now?

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem_priv.h:
```diff
@@ -557,8 +517,21 @@ static void eth_xlnx_gem##port##_irq_config(const struct device *dev)\
 /* RX/TX BD Ring initialization macro */
+#ifdef DEVICE_MMIO_IS_IN_RAM
+#define ETH_XLNX_GEM_INIT_BD_RING(port) \
+if (dev_conf->mac.phys_addr == DT_INST_REG_ADDR_BY_IDX(port, 0)) {\
```
```suggestion
if (dev == DEVICE_DT_INST_GET(port)) {\
```

just compare the device directly

maass-hamburg commented on drivers/ethernet/mdio/Kconfig.xlnx_gem:
```diff
@@ -0,0 +1,29 @@
+	bool "Xilinx GEM MDIO controller driver"
+	default y
+	depends on DT_HAS_XLNX_GEM_MDIO_ENABLED
+	depends on !QEMU_TARGET || NET_QEMU_ETHERNET
```
```suggestion
	depends on ETH_XLNX_GEM
```

ibirnbaum replied:
done

maass-hamburg commented on drivers/ethernet/mdio/mdio_xlnx_gem.c:
```diff
@@ -0,0 +1,334 @@
+
+#define XLNX_GEM_MDIO_DEV_INIT(port)\
+	DEVICE_DT_INST_DEFINE(port, &xlnx_gem_mdio_initialize, NULL, NULL,\
+			      &xlnx_gem##port##_mdio_cfg, POST_KERNEL, CONFIG_MDIO_INIT_PRIORITY,\
```
```suggestion
			      DEVICE_DT_GET(DT_INST_PARENT(port)), POST_KERNEL, CONFIG_MDIO_INIT_PRIORITY,\
```

you could put the device directly there and then
```
const struct device *mac_dev = dev->config;
```

ibirnbaum replied:
done

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -1773,12 +1560,37 @@ static void eth_xlnx_gem_handle_tx_done(const struct device *dev)
+		net_eth_carrier_on(dev_data->iface);
+	} else {
+		net_eth_carrier_off(dev_data->iface);
+	}
```
```suggestion
	}

	net_eth_carrier_set(dev_data->iface);
```

ibirnbaum replied:
done

maass-hamburg commented on drivers/ethernet/eth_xlnx_gem.c:
```diff
@@ -260,8 +227,20 @@ static void eth_xlnx_gem_iface_init(struct net_if *iface)
+	 * If PHY is available: register callback & pick up current link state immediately.
+	 * For PHY-less operation: fixed link is pre-configured, indicate link up.
+	 */
+	if (dev_conf->phy_dev != NULL) {
```
phy_dev will never be NULL, as it is required. and ``DEVICE_DT_GET(DT_INST_PHANDLE(port, phy_handle))`` will never return NULL

ibirnbaum replied:
done

maass-hamburg commented on boards/qemu/cortex_a9/Kconfig.defconfig:
```diff
@@ -42,6 +42,9 @@ configdefault NET_MGMT_EVENT_STACK_SIZE
 configdefault TEST_RANDOM_GENERATOR
 	default y if NETWORKING
 
+config ETH_PHY_DRIVER
```
```suggestion
configdefault ETH_PHY_DRIVER
```

ibirnbaum replied:
done

ibirnbaum commented:
Re-based due to conflicting changes. Twister currently fails due to unrelated ARMv7 MPU issues in testcase builds for unrelated targets.

ibirnbaum commented:
Twister has now completed successfully. I.e., this should be good to go now.
